File: editors/vscode/src/sessions.ts
import { runCli } from "./asterCli";
import {
  SessionSummary,
  TranscriptToolCall,
  TranscriptTurn,
} from "./protocol";

interface RawToolCall {
  id?: string;
  function?: { name?: string; arguments?: string };
}

interface RawEvent {
  type: string;
  role?: string;
  content?: string;
  tool_calls?: RawToolCall[];
  tool_call_id?: string;
  reasoning?: { text?: string; tokens?: number; duration_ms?: number };
}

/** Saved sessions for this repo, newest first. */
export async function listSessions(cwd: string): Promise<SessionSummary[]> {
  try {
    const { stdout, code } = await runCli(["sessions", "list", "--json"], cwd);
    if (code !== 0) {
      return [];
    }
    // The CLI already sorts newest first; reversing here put the oldest on top.
    return JSON.parse(stdout) as SessionSummary[];
  } catch {
    return [];
  }
}

/** Remove a saved session. Errors are the caller's to report. */
export async function deleteSession(cwd: string, id: string): Promise<void> {
  const { code, stderr } = await runCli(["sessions", "delete", id, "--json"], cwd);
  if (code !== 0) {
    throw new Error(stderr.trim() || `could not delete session ${id}`);
  }
}

/** Give a saved session a new name. */
export async function renameSession(cwd: string, id: string, title: string): Promise<void> {
  const { code, stderr } = await runCli(["sessions", "rename", id, title, "--json"], cwd);
  if (code !== 0) {
    throw new Error(stderr.trim() || `could not rename session ${id}`);
  }
}

/**
 * A session's transcript, rebuilt into the turns the thread renders. Assistant
 * turns keep their thinking and tool calls: replaying only `content` dropped
 * every reasoning block, every tool step, and any turn that was tool calls
 * alone, which is most of a working session.
 */
export async function loadSession(cwd: string, id: string): Promise<TranscriptTurn[]> {
  const { stdout, code } = await runCli(["sessions", "show", id, "--json"], cwd);
  if (code !== 0) {
    throw new Error(`could not load session ${id}`);
  }
  const parsed = JSON.parse(stdout) as { events?: RawEvent[] };
  const events = (parsed.events ?? []).filter((e) => e.type === "message");
  // Results arrive as their own `tool` events after the call, so index them
  // first and let each assistant turn pick up its own.
  const results = new Map<string, string>();
  for (const event of events) {
    if (event.role === "tool" && event.tool_call_id) {
      results.set(event.tool_call_id, event.content ?? "");
    }
  }

  const turns: TranscriptTurn[] = [];
  for (const event of events) {
    if (event.role === "user") {
      const content = event.content ?? "";
      // Harness steering is recorded as `system` and never lands here, but an
      // empty user turn is still nothing to draw.
      if (content.trim()) {
        turns.push({ role: "user", content });
      }
      continue;
    }
    if (event.role !== "assistant") {
      continue;
    }
    const toolCalls = (event.tool_calls ?? []).map(toTranscriptCall(results));
    const reasoning = event.reasoning?.text?.trim()
      ? {
          text: event.reasoning.text,
          tokens: event.reasoning.tokens,
          durationMs: event.reasoning.duration_ms,
        }
      : undefined;
    const content = event.content ?? "";
    // A round can be tool calls with no commentary, or thinking alone. Only a
    // turn carrying none of the three has nothing to show.
    if (!content.trim() && toolCalls.length === 0 && !reasoning) {
      continue;
    }
    turns.push({ role: "assistant", content, reasoning, toolCalls });
  }
  return turns;
}

function toTranscriptCall(results: Map<string, string>) {
  return (call: RawToolCall, index: number): TranscriptToolCall => {
    const id = call.id ?? `restored-${index}`;
    const result = results.get(id);
    return {
      id,
      name: call.function?.name ?? "unknown",
      arguments: call.function?.arguments ?? "{}",
      result,
      error: result?.startsWith("error: ") || undefined,
    };
  };
}

File: crates/aster-serve/src/sessions.rs
//! Saved sessions, and the transcript a reopened one is rebuilt from. Assistant
//! turns keep their thinking and tool calls: replaying only `content` would drop
//! every reasoning block and any turn that was tool calls alone, which is most
//! of a working session.

use serde_json::{Value, json};

use crate::cli::Cli;

pub async fn list(cli: &Cli) -> Value {
    cli.json(&["sessions", "list"], None)
        .await
        .unwrap_or_else(|_| json!([]))
}

pub async fn delete(cli: &Cli, id: &str) -> Result<(), String> {
    cli.json(&["sessions", "delete", id], None)
        .await
        .map(|_| ())
}

pub async fn rename(cli: &Cli, id: &str, title: &str) -> Result<(), String> {
    cli.json(&["sessions", "rename", id, title], None)
        .await
        .map(|_| ())
}

pub async fn load(cli: &Cli, id: &str) -> Result<Value, String> {
    let parsed = cli
        .json(&["sessions", "show", id], None)
        .await
        .map_err(|_| format!("could not load session {id}"))?;
    Ok(turns(&parsed))
}

/// Rebuild the turns the thread renders from a session's recorded events.
fn turns(session: &Value) -> Value {
    let events: Vec<&Value> = session["events"]
        .as_array()
        .map(|events| {
            events
                .iter()
                .filter(|event| event["type"] == "message")
                .collect()
        })
        .unwrap_or_default();

    // Results arrive as their own `tool` events after the call, so index them
    // first and let each assistant turn pick up its own.
    let results: std::collections::HashMap<&str, &str> = events
        .iter()
        .filter(|event| event["role"] == "tool")
        .filter_map(|event| {
            Some((
                event["tool_call_id"].as_str()?,
                event["content"].as_str().unwrap_or_default(),
            ))
        })
        .collect();

    let mut turns = Vec::new();
    for event in events {
        let content = event["content"].as_str().unwrap_or_default();
        match event["role"].as_str() {
            // Harness steering is recorded as `system` and never lands here,
            // but an empty user turn is still nothing to draw.
            Some("user") if !content.trim().is_empty() => {
                turns.push(json!({ "role": "user", "content": content }));
            }
            Some("assistant") => {
                let calls = tool_calls(&event["tool_calls"], &results);
                let reasoning = reasoning(&event["reasoning"]);
                // A round can be tool calls with no commentary, or thinking
                // alone. Only a turn carrying none of the three is empty.
                if content.trim().is_empty() && calls.is_empty() && reasoning.is_none() {
                    continue;
                }
                turns.push(json!({
                    "role": "assistant",
                    "content": content,
                    "reasoning": reasoning,
                    "toolCalls": calls,
                }));
            }
            _ => {}
        }
    }
    Value::Array(turns)
}

fn tool_calls(calls: &Value, results: &std::collections::HashMap<&str, &str>) -> Vec<Value> {
    calls
        .as_array()
        .map(|calls| {
            calls
                .iter()
                .enumerate()
                .map(|(index, call)| {
                    let id = call["id"]
                        .as_str()
                        .map(str::to_owned)
                        .unwrap_or_else(|| format!("restored-{index}"));
                    let result = results.get(id.as_str()).copied();
                    json!({
                        "id": id,
                        "name": call["function"]["name"].as_str().unwrap_or("unknown"),
                        "arguments": call["function"]["arguments"].as_str().unwrap_or("{}"),
                        "result": result,
                        "error": result.map(|r| r.starts_with("error: ")).filter(|e| *e),
                    })
                })
                .collect()
        })
        .unwrap_or_default()
}

fn reasoning(reasoning: &Value) -> Option<Value> {
    let text = reasoning["text"]
        .as_str()
        .filter(|t| !t.trim().is_empty())?;
    Some(json!({
        "text": text,
        "tokens": reasoning["tokens"],
        "durationMs": reasoning["duration_ms"],
    }))
}

#[cfg(test)]
#[path = "tests/sessions_test.rs"]
mod tests;
